## The Framingham Heart Study. Data preprocessing

### Connecting libraries and scripts

In [1]:
import pandas as pd
import numpy as np
import typing

### Settings

In [2]:
# 1. General settings
pd.set_option('display.max_columns', 100)

# 2. Warnings
import warnings
warnings.filterwarnings("ignore")

### Reduce memory usage

In [3]:
import reduce_mem_usage

### Functions

**Function creates new features based on 'PERIOD' column.**<br>

In [4]:
def new_features_by_period(
        df: pd.DataFrame, target: str,
        columns_to_create: typing.List[str]) -> pd.DataFrame:
    '''
    Function creates new columns according to 'PERIOD' column.
        
    Parameters
    ----------
    - df - pd.DataFrame.
    - target - target column, str.
    - columns_to_create - list  of new coluns wich we need to create.

    Returns
    -------
    - df - pd.DataFrame.
    '''
    for i in range(len(columns_to_create)):
        if i == 0:
            df[columns_to_create[i]] = df[target].where(df['PERIOD'] == 1, 0)
        if i == 1:
            df[columns_to_create[i]] = df[target].where(df['PERIOD'] == 2, 0)
        if i == 2:
            df[columns_to_create[i]] = df[target].where(df['PERIOD'] == 3, 0)

    return df

### Data import

In [5]:
framingham_full_df = pd.read_csv('csv/dataset_23.06.2023.csv')

### Data preprocessing

**Introduction.**<br>
Let's find out basic information about the dataset.

In [6]:
framingham_full_df.head()

RANDID  SEX  TOTCHOL  AGE  SYSBP  DIABP  CURSMOKE  CIGPDAY    BMI  \
0    2448    1    195.0   39  106.0   70.0         0      0.0  26.97   
1    2448    1    209.0   52  121.0   66.0         0      0.0    NaN   
2    6238    2    250.0   46  121.0   81.0         0      0.0  28.73   
3    6238    2    260.0   52  105.0   69.5         0      0.0  29.43   
4    6238    2    237.0   58  108.0   66.0         0      0.0  28.50   

   DIABETES  BPMEDS  HEARTRTE  GLUCOSE  educ  PREVCHD  PREVAP  PREVMI  \
0         0     0.0      80.0     77.0   4.0        0       0       0   
1         0     0.0      69.0     92.0   4.0        0       0       0   
2         0     0.0      95.0     76.0   2.0        0       0       0   
3         0     0.0      80.0     86.0   2.0        0       0       0   
4         0     0.0      80.0     71.0   2.0        0       0       0   

   PREVSTRK  PREVHYP  TIME  PERIOD  HDLC   LDLC  DEATH  ANGINA  HOSPMI  \
0         0        0     0       1   NaN    NaN      0       0       1   
1         0        0  4628       3  31.0  178.0      0       0       1   
2         0        0     0       1   NaN    NaN      0       0       0   
3         0        0  2156       2   NaN    NaN      0       0       0   
4         0        0  4344       3  54.0  141.0      0       0       0   

   MI_FCHD  ANYCHD  STROKE  CVD  HYPERTEN  TIMEAP  TIMEMI  TIMEMIFC  TIMECHD  \
0        1       1       0    1         0    8766    6438      6438     6438   
1        1       1       0    1         0    8766    6438      6438     6438   
2        0       0       0    0         0    8766    8766      8766     8766   
3        0       0       0    0         0    8766    8766      8766     8766   
4        0       0       0    0         0    8766    8766      8766     8766   

   TIMESTRK  TIMECVD  TIMEDTH  TIMEHYP  
0      8766     6438     8766     8766  
1      8766     6438     8766     8766  
2      8766     8766     8766     8766  
3      8766     8766     8766     8766  
4      8766     8766     8766     8766

In [7]:
framingham_full_df.shape

(11627, 39)

<div class="alert alert-block alert-info"> 
<b>Comments</b><br>
- Full dataset contains 11627 rows and 39 features.<br>
</div>

In [8]:
framingham_full_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11627 entries, 0 to 11626
Data columns (total 39 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   RANDID    11627 non-null  int64  
 1   SEX       11627 non-null  int64  
 2   TOTCHOL   11218 non-null  float64
 3   AGE       11627 non-null  int64  
 4   SYSBP     11627 non-null  float64
 5   DIABP     11627 non-null  float64
 6   CURSMOKE  11627 non-null  int64  
 7   CIGPDAY   11548 non-null  float64
 8   BMI       11575 non-null  float64
 9   DIABETES  11627 non-null  int64  
 10  BPMEDS    11034 non-null  float64
 11  HEARTRTE  11621 non-null  float64
 12  GLUCOSE   10187 non-null  float64
 13  educ      11332 non-null  float64
 14  PREVCHD   11627 non-null  int64  
 15  PREVAP    11627 non-null  int64  
 16  PREVMI    11627 non-null  int64  
 17  PREVSTRK  11627 non-null  int64  
 18  PREVHYP   11627 non-null  int64  
 19  TIME      11627 non-null  int64  
 20  PERIOD    11627 non-null  in

<div class="alert alert-block alert-info"> 
<b>Comments</b><br>
- All features in the dataset are numeric.<br>
- There are gaps in some features. We'll look at them later.<br>
</div>

In [9]:
framingham_full_df.describe()

RANDID           SEX       TOTCHOL           AGE         SYSBP  \
count  1.162700e+04  11627.000000  11218.000000  11627.000000  11627.000000   
mean   5.004741e+06      1.568074    241.162418     54.792810    136.324116   
std    2.900877e+06      0.495366     45.368030      9.564299     22.798625   
min    2.448000e+03      1.000000    107.000000     32.000000     83.500000   
25%    2.474378e+06      1.000000    210.000000     48.000000    120.000000   
50%    5.006008e+06      2.000000    238.000000     54.000000    132.000000   
75%    7.472730e+06      2.000000    268.000000     62.000000    149.000000   
max    9.999312e+06      2.000000    696.000000     81.000000    295.000000   

              DIABP      CURSMOKE       CIGPDAY           BMI      DIABETES  \
count  11627.000000  11627.000000  11548.000000  11575.000000  11627.000000   
mean      83.037757      0.432528      8.250346     25.877349      0.045584   
std       11.660144      0.495448     12.186888      4.102640      0.208589   
min       30.000000      0.000000      0.000000     14.430000      0.000000   
25%       75.000000      0.000000      0.000000     23.095000      0.000000   
50%       82.000000      0.000000      0.000000     25.480000      0.000000   
75%       90.000000      1.000000     20.000000     28.070000      0.000000   
max      150.000000      1.000000     90.000000     56.800000      1.000000   

             BPMEDS      HEARTRTE       GLUCOSE          educ       PREVCHD  \
count  11034.000000  11621.000000  10187.000000  11332.000000  11627.000000   
mean       0.085554     76.781516     84.124865      1.990205      0.072418   
std        0.279717     12.463359     24.993781      1.027463      0.259189   
min        0.000000     37.000000     39.000000      1.000000      0.000000   
25%        0.000000     69.000000     72.000000      1.000000      0.000000   
50%        0.000000     75.000000     80.000000      2.000000      0.000000   
75%        0.000000     85.000000     89.000000      3.000000      0.000000   
max        1.000000    220.000000    478.000000      4.000000      1.000000   

             PREVAP        PREVMI      PREVSTRK       PREVHYP          TIME  \
count  11627.000000  11627.000000  11627.000000  11627.000000  11627.000000   
mean       0.053926      0.032167      0.013073      0.459620   1957.019438   
std        0.225882      0.176450      0.113592      0.498388   1758.776927   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000   2156.000000   
75%        0.000000      0.000000      0.000000      1.000000   4252.500000   
max        1.000000      1.000000      1.000000      1.000000   4854.000000   

             PERIOD         HDLC         LDLC         DEATH        ANGINA  \
count  11627.000000  3027.000000  3026.000000  11627.000000  11627.000000   
mean       1.899286    49.364718   176.466953      0.303346      0.163585   
std        0.807407    15.626669    46.863393      0.459723      0.369914   
min        1.000000    10.000000    20.000000      0.000000      0.000000   
25%        1.000000    39.000000   145.000000      0.000000      0.000000   
50%        2.000000    48.000000   173.000000      0.000000      0.000000   
75%        3.000000    58.000000   205.000000      1.000000      0.000000   
max        3.000000   189.000000   565.000000      1.000000      1.000000   

             HOSPMI       MI_FCHD        ANYCHD        STROKE           CVD  \
count  11627.000000  11627.000000  11627.000000  11627.000000  11627.000000   
mean       0.099252      0.153780      0.271609      0.091253      0.249333   
std        0.299013      0.360753      0.444809      0.287981      0.432646   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000 

<div class="alert alert-block alert-info"> 
<b>Comments</b><br>
Features that we will consider as numeric:<br>
- TOTCHOL - Serum Total Cholesterol (mg/dL).<br>
- AGE - Age at exam (years).<br>
- SYSBP - Systolic Blood Pressure (mean of last two of three measurements) (mmHg).<br>
- DIABP - Diastolic Blood Pressure (mean of last two of three measurements) (mmHg).<br>
- CIGPDAY - Number of cigarettes smoked each day.<br>
- BMI - Body Mass Index, weight in kilograms/height meters squared.<br>
- HEARTRTE - Heart rate (Ventricular rate) in beats/min.<br>
- GLUCOSE - Casual serum glucose (mg/dL).<br>
- HDLC - High Density Lipoprotein Cholesterol (mg/dL). Available for period 3 only.<br>
- LDLC - Low Density Lipoprotein Cholesterol (mg/dL). Available for period 3 only.<br>
<br>
Features that we will consider as boolean:<br>
- SEX - Participant sex.<br>
- CURSMOKE - Current cigarette smoking at exam.<br>
- DIABETES - Diabetic according to criteria of first exam treated or first exam with casual glucose of 200 mg/dL or more.<br>
- BPMEDS - Use of anti-hypertensive medication at exam.<br>
- PREVCHD - Prevalent Coronary Heart Disease defined as pre-existing Angina Pectoris, Myocardial Infarction (hospitalized, silent or unrecognized), or Coronary Insufficiency (unstable angina).<br>
- PREVAP - Prevalent Angina Pectoris at exam.<br>
- PREVMI - Prevalent Myocardial Infarction.<br>
- PREVSTRK - Prevalent Stroke.<br>
- PREVHYP - Prevalent Hypertensive. Subject was defined as hypertensive if treated or if second exam at which mean systolic was >=140 mmHg or mean Diastolic >=90 mmHg.<br>
- DEATH - Death from any cause.<br>
- ANGINA - Angina Pectoris.<br>
- HOSPMI - Hospitalized Myocardial Infarction.<br>
- MI_FCHD - Hospitalized Myocardial Infarction or Fatal Coronary Heart Disease.<br>
- ANYCHD - Angina Pectoris, Myocardial Infarction (hospitalized and silent or unrecognized), Coronary Insufficiency (Unstable Angina), or Fatal Coronary Heart Disease.<br>
- STROKE - Atherothrombotic Infarction, Cerebral Embolism, Intracerebral Hemorrhage, or Subarachnoid Hemorrhage or Fatal Cerebrovascular Disease.<br>
- CVD - Myocardial Infarction (hospitalized and silent or unrecognized), Fatal Coronary Heart Disease, Atherothrombotic Infarction, Cerebral Embolism, Intracerebral Hemorrhage, or Subarachnoid Hemorrhage or Fatal Cerebrovascular Disease.<br>
- HYPERTEN - Hypertensive. Defined as the first exam treated for high blood pressure or second exam in which either Systolic is 140 mmHg or Diastolic 90 mmHg.<br>
<br>
Features that we will consider as categorical:<br>
- educ - Attained education.<br>
<br>
Features that can help in EDA analysis or Feature Engeneering, but we will not use them in modeling:<br>
- RANDID - Unique identification number for each participant.<br>
- TIME - Number of days since baseline exam.<br>
- PERIOD - Examination cycle.<br>
- TIMEAP - Number of days from baseline exam to first Angina during the followup or number of days from baseline to censor date. Censor date may be end of followup, death or last known contact date if subject is lost to followup.<br>
- TIMEMI - Defined as above for the first HOSPMI event during followup.<br>
- TIMEMIFC - Defined as above for the first MI_FCHD event during followup.<br>
- TIMECHD - Defined as above for the first ANYCHD event during followup.<br>
- TIMESTRK - Defined as above for the first STROKE event during followup.<br>
- TIMECVD - Defined as above for the first CVD event during followup.<br>
- TIMEDTH - Number of days from baseline exam to death if occurring during followup or number of days from baseline to censor date. Censor date may be end of followup, or last known contact date if subject is lost to followup.<br>
- TIMEHYP - Defined as above for the first HYPERTEN event during followup.<br>

<div class="alert alert-block alert-info"> 
<b>Comments</b><br>
- There are no visible logical violations in the data.<br>
</div>

**Gaps in data.**<br>
Let's handle the gaps in the data.

In [10]:
framingham_full_df.isna().sum()

RANDID         0
SEX            0
TOTCHOL      409
AGE            0
SYSBP          0
DIABP          0
CURSMOKE       0
CIGPDAY       79
BMI           52
DIABETES       0
BPMEDS       593
HEARTRTE       6
GLUCOSE     1440
educ         295
PREVCHD        0
PREVAP         0
PREVMI         0
PREVSTRK       0
PREVHYP        0
TIME           0
PERIOD         0
HDLC        8600
LDLC        8601
DEATH          0
ANGINA         0
HOSPMI         0
MI_FCHD        0
ANYCHD         0
STROKE         0
CVD            0
HYPERTEN       0
TIMEAP         0
TIMEMI         0
TIMEMIFC       0
TIMECHD        0
TIMESTRK       0
TIMECVD        0
TIMEDTH        0
TIMEHYP        0
dtype: int64

In [11]:
(framingham_full_df.isna().sum() / len(framingham_full_df)).round(4) * 100

RANDID       0.00
SEX          0.00
TOTCHOL      3.52
AGE          0.00
SYSBP        0.00
DIABP        0.00
CURSMOKE     0.00
CIGPDAY      0.68
BMI          0.45
DIABETES     0.00
BPMEDS       5.10
HEARTRTE     0.05
GLUCOSE     12.38
educ         2.54
PREVCHD      0.00
PREVAP       0.00
PREVMI       0.00
PREVSTRK     0.00
PREVHYP      0.00
TIME         0.00
PERIOD       0.00
HDLC        73.97
LDLC        73.97
DEATH        0.00
ANGINA       0.00
HOSPMI       0.00
MI_FCHD      0.00
ANYCHD       0.00
STROKE       0.00
CVD          0.00
HYPERTEN     0.00
TIMEAP       0.00
TIMEMI       0.00
TIMEMIFC     0.00
TIMECHD      0.00
TIMESTRK     0.00
TIMECVD      0.00
TIMEDTH      0.00
TIMEHYP      0.00
dtype: float64

<div class="alert alert-block alert-info"> 
<b>Comments</b><br>
There are gaps in the following features:<br>
- TOTCHOL - 3.52%.<br>
- CIGPDAY - 0.68%.<br>
- BMI - 0.45%.<br>
- BPMEDS - 5.10%.<br>
- HEARTRTE - 0.05%.<br>
- GLUCOSE - 12.38%.<br>
- educ - 2.54%.<br>
- HDLC - 73.97%.<br>
- LDLC - 73.97%.<br>
</div>

In [12]:
len(framingham_full_df[framingham_full_df.duplicated(['RANDID'])])

7193

<div class="alert alert-block alert-info"> 
<b>Comments</b><br>
- There are duplicates in the dataset with the same IDs, but they contain different data.<br>
- Let's try to fill in the gaps with data from the same IDs.<br>
</div>

In [13]:
framingham_full_df.sort_values(by='RANDID', inplace=True)
framingham_full_df

RANDID  SEX  TOTCHOL  AGE  SYSBP  DIABP  CURSMOKE  CIGPDAY    BMI  \
0         2448    1    195.0   39  106.0   70.0         0      0.0  26.97   
1         2448    1    209.0   52  121.0   66.0         0      0.0    NaN   
2         6238    2    250.0   46  121.0   81.0         0      0.0  28.73   
3         6238    2    260.0   52  105.0   69.5         0      0.0  29.43   
4         6238    2    237.0   58  108.0   66.0         0      0.0  28.50   
...        ...  ...      ...  ...    ...    ...       ...      ...    ...   
11621  9998212    1    185.0   40  141.0   98.0         0      0.0  25.60   
11623  9998212    1    153.0   52  143.0   89.0         0      0.0  25.74   
11625  9999312    2    240.0   46  138.0   79.0         1     20.0  26.39   
11624  9999312    2    196.0   39  133.0   86.0         1     30.0  20.91   
11626  9999312    2      NaN   50  147.0   96.0         1     10.0  24.19   

       DIABETES  BPMEDS  HEARTRTE  GLUCOSE  educ  PREVCHD  PREVAP  PREVMI  \
0             0     0.0      80.0     77.0   4.0        0       0       0   
1             0     0.0      69.0     92.0   4.0        0       0       0   
2             0     0.0      95.0     76.0   2.0        0       0       0   
3             0     0.0      80.0     86.0   2.0        0       0       0   
4             0     0.0      80.0     71.0   2.0        0       0       0   
...         ...     ...       ...      ...   ...      ...     ...     ...   
11621         0     0.0      67.0     72.0   3.0        0       0       0   
11623         0     0.0      65.0     72.0   3.0        0       0       0   
11625         0     0.0      90.0     83.0   3.0        0       0       0   
11624         0     0.0      85.0     80.0   3.0        0       0       0   
11626         0     0.0      94.0      NaN   3.0        0       0       0   

       PREVSTRK  PREVHYP  TIME  PERIOD  HDLC   LDLC  DEATH  ANGINA  HOSPMI  \
0             0        0     0       1   NaN    NaN      0       0       1   
1             0        0  4628       3  31.0  178.0      0       0       1   
2             0        0     0       1   NaN    NaN      0       0       0   
3             0        0  2156       2   NaN    NaN      0       0       0   
4             0        0  4344       3  54.0  141.0      0       0       0   
...         ...      ...   ...     ...   ...    ...    ...     ...     ...   
11621         0        1     0       1   NaN    NaN      0       0       0   
11623         0        1  4538       3  30.0  123.0      0       0       0   
11625         0        0  2390       2   NaN    NaN      0       0       0   
11624         0        0     0       1   NaN    NaN      0       0       0   
11626         0        1  4201       3   NaN    NaN      0       0       0   

       MI_FCHD  ANYCHD  STROKE  CVD  HYPERTEN  TIMEAP  TIMEMI  TIMEMIFC  \
0            1       1       0    1         0    8766    6438      6438   
1            1       1       0    1         0    8766    6438      6438   
2            0       0       0    0         0    8766    8766      8766   
3            0       0       0    0         0    8766    8766      8766   
4            0       0       0    0         0    8766    8766      8766   
...        ...     ...     ...  ...       ...     ...     ...       ...   
11621        0       0       0    0         1    8766    8766      8766   
11623        0       0       0    0         1    8766    8766      8766   
11625        0       0       0    0         1    8766    8766      8766   
11624        0       0       0    0         1    8766    8766      8766   
11626        0       0       0    0         1    8766    8766      8766   

       TIMECHD  TIMESTRK  TIMECVD  TIMEDTH  TIMEHYP  
0         6438      8766     6438     8766     8766  
1         6438      8766     6438     8766     8766  
2         8766      8766     8766     8766     8766  
3         8766      8766     8766     8766     8766  
4         8766      8766     8766     8766     876

In [14]:
gaps_columns = [
    'TOTCHOL', 'CIGPDAY', 'BMI',
    'BPMEDS', 'HEARTRTE', 'GLUCOSE', 
    'educ', 'HDLC', 'LDLC'
]

for column in gaps_columns:
    framingham_full_df[column] = framingham_full_df.groupby(
        'RANDID')[column].ffill()

<div class="alert alert-block alert-info"> 
<b>Comments</b><br>
The number of gaps in the data has decreased. But still, gaps remained.<br>
We will fill the remaining gaps in the data with the median or mode,<br>
depending on the attribute, grouping values by 'AGE'.<br>
<br>
There are gaps in the following features:<br>
- TOTCHOL - 0.92% - median.<br>
- CIGPDAY - 0.24% - median.<br>
- BMI - 0.23% - median.<br>
- BPMEDS - 0.99% - mode.<br>
- HEARTRTE - 0.01% - median.<br>
- GLUCOSE - 4.64% - median.<br>
- educ - 2.54% - mode.<br>
- HDLC - 64.05% - median.<br>
- LDLC - 64.06% - median.<br>
<br>
Gaps in the data are small, so filling will not greatly affect the distribution of variables.<br>
</div>

In [15]:
gaps_columns_median = [
    'TOTCHOL', 'CIGPDAY', 'BMI',
    'HEARTRTE', 'GLUCOSE', 'HDLC', 'LDLC'
]
gaps_columns_mode = ['BPMEDS', 'educ']

for column in gaps_columns_median:
    framingham_full_df[column].fillna(framingham_full_df.groupby('AGE')
                                      [column].transform(lambda x: x.median()),
                                      inplace=True)

for column in gaps_columns_mode:
    framingham_full_df[column].fillna(framingham_full_df.groupby(
        'AGE')[column].transform(lambda x: x.mode()[0]), inplace=True)

<div class="alert alert-block alert-info"> 
<b>Comments</b><br>
- There is one gap left in the features 'HDLC' and 'LDLC'.<br>
- Let's fill them with the median according to the feature.<br>
</div>

In [16]:
framingham_full_df['HDLC'].fillna(framingham_full_df['HDLC'].median(),
                                  inplace=True)
framingham_full_df['LDLC'].fillna(framingham_full_df['LDLC'].median(),
                                  inplace=True)

In [17]:
(framingham_full_df.isna().sum() / len(framingham_full_df)).round(4) * 100

RANDID      0.0
SEX         0.0
TOTCHOL     0.0
AGE         0.0
SYSBP       0.0
DIABP       0.0
CURSMOKE    0.0
CIGPDAY     0.0
BMI         0.0
DIABETES    0.0
BPMEDS      0.0
HEARTRTE    0.0
GLUCOSE     0.0
educ        0.0
PREVCHD     0.0
PREVAP      0.0
PREVMI      0.0
PREVSTRK    0.0
PREVHYP     0.0
TIME        0.0
PERIOD      0.0
HDLC        0.0
LDLC        0.0
DEATH       0.0
ANGINA      0.0
HOSPMI      0.0
MI_FCHD     0.0
ANYCHD      0.0
STROKE      0.0
CVD         0.0
HYPERTEN    0.0
TIMEAP      0.0
TIMEMI      0.0
TIMEMIFC    0.0
TIMECHD     0.0
TIMESTRK    0.0
TIMECVD     0.0
TIMEDTH     0.0
TIMEHYP     0.0
dtype: float64

<div class="alert alert-block alert-info"> 
<b>Comments</b><br>
- All data gaps have been eliminated.<br>
</div>

**Feature selection.**<br>
Target variable - CVD.
Cardiovascular disease - is the term for all types of diseases that affect the heart or blood vessels, including coronary heart disease (clogged arteries), which can cause heart attacks, stroke, heart failure, and peripheral artery disease.<br>
<br>
Our task is to predict the occurrence of CVD based on the data provided by the patients.<br>
<br>
So let’s delete columns not directly related to CVD.<br>

In [18]:
framingham_full_df

RANDID  SEX  TOTCHOL  AGE  SYSBP  DIABP  CURSMOKE  CIGPDAY    BMI  \
0         2448    1    195.0   39  106.0   70.0         0      0.0  26.97   
1         2448    1    209.0   52  121.0   66.0         0      0.0  26.97   
2         6238    2    250.0   46  121.0   81.0         0      0.0  28.73   
3         6238    2    260.0   52  105.0   69.5         0      0.0  29.43   
4         6238    2    237.0   58  108.0   66.0         0      0.0  28.50   
...        ...  ...      ...  ...    ...    ...       ...      ...    ...   
11621  9998212    1    185.0   40  141.0   98.0         0      0.0  25.60   
11623  9998212    1    153.0   52  143.0   89.0         0      0.0  25.74   
11625  9999312    2    240.0   46  138.0   79.0         1     20.0  26.39   
11624  9999312    2    196.0   39  133.0   86.0         1     30.0  20.91   
11626  9999312    2    196.0   50  147.0   96.0         1     10.0  24.19   

       DIABETES  BPMEDS  HEARTRTE  GLUCOSE  educ  PREVCHD  PREVAP  PREVMI  \
0             0     0.0      80.0     77.0   4.0        0       0       0   
1             0     0.0      69.0     92.0   4.0        0       0       0   
2             0     0.0      95.0     76.0   2.0        0       0       0   
3             0     0.0      80.0     86.0   2.0        0       0       0   
4             0     0.0      80.0     71.0   2.0        0       0       0   
...         ...     ...       ...      ...   ...      ...     ...     ...   
11621         0     0.0      67.0     72.0   3.0        0       0       0   
11623         0     0.0      65.0     72.0   3.0        0       0       0   
11625         0     0.0      90.0     83.0   3.0        0       0       0   
11624         0     0.0      85.0     80.0   3.0        0       0       0   
11626         0     0.0      94.0     80.0   3.0        0       0       0   

       PREVSTRK  PREVHYP  TIME  PERIOD  HDLC   LDLC  DEATH  ANGINA  HOSPMI  \
0             0        0     0       1  48.0  168.0      0       0       1   
1             0        0  4628       3  31.0  178.0      0       0       1   
2             0        0     0       1  47.0  165.0      0       0       0   
3             0        0  2156       2  46.0  168.0      0       0       0   
4             0        0  4344       3  54.0  141.0      0       0       0   
...         ...      ...   ...     ...   ...    ...    ...     ...     ...   
11621         0        1     0       1  48.0  164.0      0       0       0   
11623         0        1  4538       3  30.0  123.0      0       0       0   
11625         0        0  2390       2  47.0  165.0      0       0       0   
11624         0        0     0       1  48.0  168.0      0       0       0   
11626         0        1  4201       3  48.0  170.0      0       0       0   

       MI_FCHD  ANYCHD  STROKE  CVD  HYPERTEN  TIMEAP  TIMEMI  TIMEMIFC  \
0            1       1       0    1         0    8766    6438      6438   
1            1       1       0    1         0    8766    6438      6438   
2            0       0       0    0         0    8766    8766      8766   
3            0       0       0    0         0    8766    8766      8766   
4            0       0       0    0         0    8766    8766      8766   
...        ...     ...     ...  ...       ...     ...     ...       ...   
11621        0       0       0    0         1    8766    8766      8766   
11623        0       0       0    0         1    8766    8766      8766   
11625        0       0       0    0         1    8766    8766      8766   
11624        0       0       0    0         1    8766    8766      8766   
11626        0       0       0    0         1    8766    8766      8766   

       TIMECHD  TIMESTRK  TIMECVD  TIMEDTH  TIMEHYP  
0         6438      8766     6438     8766     8766  
1         6438      8766     6438     8766     8766  
2         8766      8766     8766     8766     8766  
3         8766      8766     8766     8766     8766  
4         8766      8766     8766     8766     876

In [19]:
framingham_filtered_df = framingham_full_df.drop([
    'DEATH', 'ANGINA', 'HOSPMI', 'MI_FCHD', 'ANYCHD', 'STROKE', 'HYPERTEN',
    'TIMEAP', 'TIMEMI', 'TIMEMIFC', 'TIMECHD', 'TIMESTRK', 'TIMEDTH', 'TIMEHYP',
    ], axis=1)

**Feature Engeneering.**<br>
**1 step.**<br>
The study contains 3 observation periods. Therefore, there can be up to 3 records per patient.<br>
This is illustrated more clearly in the example of patient ID 95148 below.<br>
<br>
In the base dataset, if the CVD occurred on any day of the study, then in each period the 'CVD' column will contain 1.<br>
<br>
We need to make sure that 1 stands only in the period when the CVD actually happened.<br>
<br>
Next, in each period we found the lower and upper time limits based on the 'TIME' column.<br>
<br>
In the new column 'CVD_BY_PERIOD', we put 1 where the time of appearance of the CVD falls within the boundaries of the period.
<br>

**Example.**<br>

In [20]:
framingham_filtered_df[framingham_full_df['RANDID'] == 95148].sort_index()

RANDID  SEX  TOTCHOL  AGE  SYSBP  DIABP  CURSMOKE  CIGPDAY    BMI  \
112   95148    2    240.0   52  148.0   92.0         0      0.0  25.09   
113   95148    2    268.0   58  135.0   75.0         0      0.0  26.55   
114   95148    2    240.0   64  154.0   98.0         0      0.0  26.91   

     DIABETES  BPMEDS  HEARTRTE  GLUCOSE  educ  PREVCHD  PREVAP  PREVMI  \
112         0     1.0      70.0     79.0   1.0        0       0       0   
113         0     1.0      75.0    105.0   1.0        0       0       0   
114         0     0.0      70.0     81.0   1.0        1       1       1   

     PREVSTRK  PREVHYP  TIME  PERIOD  HDLC   LDLC  CVD  TIMECVD  
112         0        1     0       1  46.0  168.0    1     3607  
113         0        1  2128       2  46.0  168.0    1     3607  
114         0        1  4192       3  46.0  168.0    1     3607

In [21]:
framingham_filtered_df_1 = framingham_filtered_df[framingham_full_df['PERIOD']
                                                  == 1].sort_index()
framingham_filtered_df_2 = framingham_filtered_df[framingham_full_df['PERIOD']
                                                  == 2].sort_index()
framingham_filtered_df_3 = framingham_filtered_df[framingham_full_df['PERIOD']
                                                  == 3].sort_index()

<div class="alert alert-block alert-info"> 
<b>Comments</b><br>
- We divided the dataset into 3 research periods to make it easier to process the data.<br>
</div>

In [22]:
# new_columns = ['CVD_BY_PERIOD', 'ANGINA_BY_PERIOD', 'HOSPMI_BY_PERIOD', 'MI_FCHD_BY_PERIOD', 'ANYCHD_BY_PERIOD', 'STROKE_BY_PERIOD', 'HYPERTEN_BY_PERIOD']
# columns = ['CVD', 'ANGINA', 'HOSPMI', 'MI_FCHD', 'ANYCHD', 'STROKE', 'HYPERTEN']
# time_columns = ['TIMECVD', 'TIMEAP', 'TIMEMI', 'TIMEMIFC', 'TIMECHD', 'TIMESTRK', 'TIMEHYP']

In [23]:
maximum = min(framingham_filtered_df_2['TIME'].to_list())

framingham_filtered_df_1['CVD_BY_PERIOD'] = np.where(
    (framingham_filtered_df_1['CVD'] == 1) &
    ((framingham_filtered_df_1['TIMECVD'] == 0) |
     (framingham_filtered_df_1['TIMECVD'] < maximum)),
    1,
    0,
)

# for i in range(len(columns)):
#     framingham_filtered_df_1[new_columns[i]] = np.where(
#         (framingham_filtered_df_1[columns[i]] == 1) &
#         ((framingham_filtered_df_1[time_columns[i]] == 0) |
#          (framingham_filtered_df_1[time_columns[i]] < maximum)),
#         1,
#         0,
#     )

In [24]:
minimum = min(framingham_filtered_df_2['TIME'].to_list())
maximum = min(framingham_filtered_df_3['TIME'].to_list())

framingham_filtered_df_2['CVD_BY_PERIOD'] = np.where(
    (framingham_filtered_df_2['CVD'] == 1) &
    ((minimum <= framingham_filtered_df_2['TIMECVD']) &
     (framingham_filtered_df_2['TIMECVD'] < maximum)), 1, 0)

# for i in range(len(columns)):
#     framingham_filtered_df_2[new_columns[i]] = np.where(
#         (framingham_filtered_df_2[columns[i]] == 1) &
#         ((minimum <= framingham_filtered_df_2[time_columns[i]]) &
#          (framingham_filtered_df_2[time_columns[i]] < maximum)), 1, 0)

In [25]:
minimum = min(framingham_filtered_df_3['TIME'].to_list())

framingham_filtered_df_3['CVD_BY_PERIOD'] = np.where(
    (framingham_filtered_df_3['CVD'] == 1) &
    (framingham_filtered_df_3['TIMECVD'] >= minimum), 1, 0)

# for i in range(len(columns)):
#     framingham_filtered_df_3[new_columns[i]] = np.where(
#         (framingham_filtered_df_3[columns[i]] == 1) &
#         (framingham_filtered_df_3[time_columns[i]] >= minimum), 1, 0)

<div class="alert alert-block alert-info"> 
<b>Comments</b><br>
- We determined the time frame for each period and put it in a new column 'CVD_BY_PERIOD' 1 where the CVD actually appeared.<br>
</div>

In [26]:
framingham_merged_df = pd.concat([
    framingham_filtered_df_1, framingham_filtered_df_2,
    framingham_filtered_df_3],
    ignore_index=True)

<div class="alert alert-block alert-info"> 
<b>Comments</b><br>
- Merged 3 periods into one.<br>
</div>

In [27]:
framingham_merged_df[framingham_merged_df['RANDID'] == 95148].sort_index()

RANDID  SEX  TOTCHOL  AGE  SYSBP  DIABP  CURSMOKE  CIGPDAY    BMI  \
44     95148    2    240.0   52  148.0   92.0         0      0.0  25.09   
4470   95148    2    268.0   58  135.0   75.0         0      0.0  26.55   
8396   95148    2    240.0   64  154.0   98.0         0      0.0  26.91   

      DIABETES  BPMEDS  HEARTRTE  GLUCOSE  educ  PREVCHD  PREVAP  PREVMI  \
44           0     1.0      70.0     79.0   1.0        0       0       0   
4470         0     1.0      75.0    105.0   1.0        0       0       0   
8396         0     0.0      70.0     81.0   1.0        1       1       1   

      PREVSTRK  PREVHYP  TIME  PERIOD  HDLC   LDLC  CVD  TIMECVD  \
44           0        1     0       1  46.0  168.0    1     3607   
4470         0        1  2128       2  46.0  168.0    1     3607   
8396         0        1  4192       3  46.0  168.0    1     3607   

      CVD_BY_PERIOD  
44                0  
4470              1  
8396              0

**Feature Engeneering.**<br>
**2 step.**<br>
Let's create additional columns that will be responsible for the values in a particular period.<br>

In [28]:
data = {
    'AGE': ['AGE_1', 'AGE_2', 'AGE_3'],
    'CURSMOKE': ['CURSMOKE_1', 'CURSMOKE_2', 'CURSMOKE_3'],
    'DIABETES': ['DIABETES_1', 'DIABETES_2', 'DIABETES_3'],
    'BPMEDS': ['BPMEDS_1', 'BPMEDS_2', 'BPMEDS_3'],
    'PREVCHD': ['PREVCHD_1', 'PREVCHD_2', 'PREVCHD_3'],
    'PREVAP': ['PREVAP_1', 'PREVAP_2', 'PREVAP_3'],
    'PREVMI': ['PREVMI_1', 'PREVMI_2', 'PREVMI_3'],
    'PREVSTRK': ['PREVSTRK_1', 'PREVSTRK_2', 'PREVSTRK_3'],
    'PREVHYP': ['PREVHYP_1', 'PREVHYP_2', 'PREVHYP_3'],
    'CVD_BY_PERIOD': ['CVD_BY_PERIOD_1', 'CVD_BY_PERIOD_2', 'CVD_BY_PERIOD_3']
    # 'ANGINA_BY_PERIOD': ['ANGINA_BY_PERIOD_1', 'ANGINA_BY_PERIOD_2', 'ANGINA_BY_PERIOD_3'],
    # 'HOSPMI_BY_PERIOD': ['HOSPMI_BY_PERIOD_1', 'HOSPMI_BY_PERIOD_2', 'HOSPMI_BY_PERIOD_3'],
    # 'MI_FCHD_BY_PERIOD': ['MI_FCHD_BY_PERIOD_1', 'MI_FCHD_BY_PERIOD_2', 'MI_FCHD_BY_PERIOD_3'],
    # 'ANYCHD_BY_PERIOD': ['ANYCHD_BY_PERIOD_1', 'ANYCHD_BY_PERIOD_2', 'ANYCHD_BY_PERIOD_3'],
    # 'STROKE_BY_PERIOD': ['STROKE_BY_PERIOD_1', 'STROKE_BY_PERIOD_2', 'STROKE_BY_PERIOD_3'],
    # 'HYPERTEN_BY_PERIOD': ['HYPERTEN_BY_PERIOD_1', 'HYPERTEN_BY_PERIOD_2', 'HYPERTEN_BY_PERIOD_3'],
}

for key, value in data.items():
    framingham_merged_df = new_features_by_period(framingham_merged_df,
                                                  key,
                                                  value)

framingham_merged_df[framingham_merged_df['RANDID'] == 95148].sort_index()

RANDID  SEX  TOTCHOL  AGE  SYSBP  DIABP  CURSMOKE  CIGPDAY    BMI  \
44     95148    2    240.0   52  148.0   92.0         0      0.0  25.09   
4470   95148    2    268.0   58  135.0   75.0         0      0.0  26.55   
8396   95148    2    240.0   64  154.0   98.0         0      0.0  26.91   

      DIABETES  BPMEDS  HEARTRTE  GLUCOSE  educ  PREVCHD  PREVAP  PREVMI  \
44           0     1.0      70.0     79.0   1.0        0       0       0   
4470         0     1.0      75.0    105.0   1.0        0       0       0   
8396         0     0.0      70.0     81.0   1.0        1       1       1   

      PREVSTRK  PREVHYP  TIME  PERIOD  HDLC   LDLC  CVD  TIMECVD  \
44           0        1     0       1  46.0  168.0    1     3607   
4470         0        1  2128       2  46.0  168.0    1     3607   
8396         0        1  4192       3  46.0  168.0    1     3607   

      CVD_BY_PERIOD  AGE_1  AGE_2  AGE_3  CURSMOKE_1  CURSMOKE_2  CURSMOKE_3  \
44                0     52      0      0           0           0           0   
4470              1      0     58      0           0           0           0   
8396              0      0      0     64           0           0           0   

      DIABETES_1  DIABETES_2  DIABETES_3  BPMEDS_1  BPMEDS_2  BPMEDS_3  \
44             0           0           0       1.0       0.0       0.0   
4470           0           0           0       0.0       1.0       0.0   
8396           0           0           0       0.0       0.0       0.0   

      PREVCHD_1  PREVCHD_2  PREVCHD_3  PREVAP_1  PREVAP_2  PREVAP_3  PREVMI_1  \
44            0          0          0         0         0         0         0   
4470          0          0          0         0         0         0         0   
8396          0          0          1         0         0         1         0   

      PREVMI_2  PREVMI_3  PREVSTRK_1  PREVSTRK_2  PREVSTRK_3  PREVHYP_1  \
44           0         0           0           0           0          1   
4470         0         0           0           0           0          0   
8396         0         1           0           0           0          0   

      PREVHYP_2  PREVHYP_3  CVD_BY_PERIOD_1  CVD_BY_PERIOD_2  CVD_BY_PERIOD_3  
44            0          0                0                0                0  
4470          1          0                0                1                0  
8396          0          1                0                0                0

**Feature Engeneering.**<br>
**3 step.**<br>
Let's convert the 'SEX' column to binary.<br>
1 = Men.<br>
2 = Women.<br>
<br>
Convert to:<br>
0 = Men.<br>
1 = Women.<br>

In [29]:
dict_sex = {
    1: 0,
    2: 1
}

In [30]:
framingham_merged_df['SEX'] = framingham_merged_df['SEX'].map(dict_sex)

framingham_merged_df.head()

RANDID  SEX  TOTCHOL  AGE  SYSBP  DIABP  CURSMOKE  CIGPDAY    BMI  \
0    2448    0    195.0   39  106.0   70.0         0      0.0  26.97   
1    6238    1    250.0   46  121.0   81.0         0      0.0  28.73   
2    9428    0    245.0   48  127.5   80.0         1     20.0  25.34   
3   10552    1    225.0   61  150.0   95.0         1     30.0  28.58   
4   11252    1    285.0   46  130.0   84.0         1     23.0  23.10   

   DIABETES  BPMEDS  HEARTRTE  GLUCOSE  educ  PREVCHD  PREVAP  PREVMI  \
0         0     0.0      80.0     77.0   4.0        0       0       0   
1         0     0.0      95.0     76.0   2.0        0       0       0   
2         0     0.0      75.0     70.0   1.0        0       0       0   
3         0     0.0      65.0    103.0   3.0        0       0       0   
4         0     0.0      85.0     85.0   3.0        0       0       0   

   PREVSTRK  PREVHYP  TIME  PERIOD  HDLC   LDLC  CVD  TIMECVD  CVD_BY_PERIOD  \
0         0        0     0       1  48.0  168.0    1     6438              0   
1         0        0     0       1  47.0  165.0    0     8766              0   
2         0        0     0       1  48.0  176.0    0     8766              0   
3         0        1     0       1  46.0  179.0    1     2089              0   
4         0        0     0       1  47.0  165.0    0     8766              0   

   AGE_1  AGE_2  AGE_3  CURSMOKE_1  CURSMOKE_2  CURSMOKE_3  DIABETES_1  \
0     39      0      0           0           0           0           0   
1     46      0      0           0           0           0           0   
2     48      0      0           1           0           0           0   
3     61      0      0           1           0           0           0   
4     46      0      0           1           0           0           0   

   DIABETES_2  DIABETES_3  BPMEDS_1  BPMEDS_2  BPMEDS_3  PREVCHD_1  PREVCHD_2  \
0           0           0       0.0       0.0       0.0          0          0   
1           0           0       0.0       0.0       0.0          0          0   
2           0           0       0.0       0.0       0.0          0          0   
3           0           0       0.0       0.0       0.0          0          0   
4           0           0       0.0       0.0       0.0          0          0   

   PREVCHD_3  PREVAP_1  PREVAP_2  PREVAP_3  PREVMI_1  PREVMI_2  PREVMI_3  \
0          0         0         0         0         0         0         0   
1          0         0         0         0         0         0         0   
2          0         0         0         0         0         0         0   
3          0         0         0         0         0         0         0   
4          0         0         0         0         0         0         0   

   PREVSTRK_1  PREVSTRK_2  PREVSTRK_3  PREVHYP_1  PREVHYP_2  PREVHYP_3  \
0           0           0           0          0          0          0   
1           0           0           0          0          0          0   
2           0           0           0          0          0          0   
3           0           0           0          1          0          0   
4           0           0           0          0          0          0   

   CVD_BY_PERIOD_1  CVD_BY_PERIOD_2  CVD_BY_PERIOD_3  
0                0                0                0  
1                0                0                0  
2                0                0                0  
3                0                0                0  
4                0                0                0

**Grouping.**<br>
The study contains 3 observation periods. Therefore, there can be up to 3 records per patient.<br>
The study contains 4434 unique patients.<br>
We can group features that we will consider as numeric by median value.<br>
And we can group features that we will consider as boolean and categorical by max value.<br>

In [31]:
framingham_grouped_1 = framingham_merged_df.groupby(['RANDID'],
                                                    as_index=False)[[
                                                        'TOTCHOL', 'AGE',
                                                        'SYSBP', 'DIABP',
                                                        'CIGPDAY', 'BMI',
                                                        'HEARTRTE', 'GLUCOSE',
                                                        'HDLC', 'LDLC'
                                                    ]].median()

framingham_grouped_2 = framingham_merged_df.groupby(
    ['RANDID'], as_index=False)[[
        'SEX', 'educ', 'CVD', 'AGE_1', 'AGE_2', 'AGE_3',
        'CURSMOKE_1', 'CURSMOKE_2', 'CURSMOKE_3',
        'DIABETES_1', 'DIABETES_2', 'DIABETES_3',
        'BPMEDS_1', 'BPMEDS_2', 'BPMEDS_3',
        'PREVCHD_1', 'PREVCHD_2', 'PREVCHD_3',
        'PREVAP_1', 'PREVAP_2', 'PREVAP_3', 
        'PREVMI_1', 'PREVMI_2', 'PREVMI_3',
        'PREVSTRK_1', 'PREVSTRK_2', 'PREVSTRK_3',
        'PREVHYP_1', 'PREVHYP_2', 'PREVHYP_3',
        'CVD_BY_PERIOD_1', 'CVD_BY_PERIOD_2', 'CVD_BY_PERIOD_3'
        # 'ANGINA_BY_PERIOD_1', 'ANGINA_BY_PERIOD_2', 'ANGINA_BY_PERIOD_3',
        # 'HOSPMI_BY_PERIOD_1', 'HOSPMI_BY_PERIOD_2', 'HOSPMI_BY_PERIOD_3',
        # 'MI_FCHD_BY_PERIOD_1', 'MI_FCHD_BY_PERIOD_2', 'MI_FCHD_BY_PERIOD_3',
        # 'ANYCHD_BY_PERIOD_1', 'ANYCHD_BY_PERIOD_2', 'ANYCHD_BY_PERIOD_3',
        # 'STROKE_BY_PERIOD_1', 'STROKE_BY_PERIOD_2', 'STROKE_BY_PERIOD_3',
        # 'HYPERTEN_BY_PERIOD_1', 'HYPERTEN_BY_PERIOD_2', 'HYPERTEN_BY_PERIOD_3'
    ]].max()

In [32]:
framingham_grouped_1[framingham_grouped_1['RANDID'] == 95148]

RANDID  TOTCHOL   AGE  SYSBP  DIABP  CIGPDAY    BMI  HEARTRTE  GLUCOSE  \
44   95148    240.0  58.0  148.0   92.0      0.0  26.55      70.0     81.0   

    HDLC   LDLC  
44  46.0  168.0

In [33]:
framingham_grouped_2[framingham_grouped_1['RANDID'] == 95148]

RANDID  SEX  educ  CVD  AGE_1  AGE_2  AGE_3  CURSMOKE_1  CURSMOKE_2  \
44   95148    1   1.0    1     52     58     64           0           0   

    CURSMOKE_3  DIABETES_1  DIABETES_2  DIABETES_3  BPMEDS_1  BPMEDS_2  \
44           0           0           0           0       1.0       1.0   

    BPMEDS_3  PREVCHD_1  PREVCHD_2  PREVCHD_3  PREVAP_1  PREVAP_2  PREVAP_3  \
44       0.0          0          0          1         0         0         1   

    PREVMI_1  PREVMI_2  PREVMI_3  PREVSTRK_1  PREVSTRK_2  PREVSTRK_3  \
44         0         0         1           0           0           0   

    PREVHYP_1  PREVHYP_2  PREVHYP_3  CVD_BY_PERIOD_1  CVD_BY_PERIOD_2  \
44          1          1          1                0                1   

    CVD_BY_PERIOD_3  
44                0

In [34]:
framingham_final_merged = framingham_grouped_1.merge(framingham_grouped_2,
                                                     on='RANDID',
                                                     how='left')
framingham_final_merged

RANDID  TOTCHOL   AGE   SYSBP  DIABP  CIGPDAY    BMI  HEARTRTE  \
0        2448    202.0  45.5  113.50   68.0      0.0  26.97      74.5   
1        6238    250.0  52.0  108.00   69.5      0.0  28.73      80.0   
2        9428    264.0  51.0  134.25   84.5     25.0  25.34      75.0   
3       10552    228.5  64.0  166.50  102.0     25.0  29.38      62.5   
4       11252    285.0  51.0  130.00   84.0     30.0  23.48      85.0   
...       ...      ...   ...     ...    ...      ...    ...       ...   
4429  9990894    248.0  54.0  145.00   89.0     20.0  22.00      80.0   
4430  9993179    236.0  50.0  131.00   87.0     25.0  21.22      86.0   
4431  9995546    267.0  55.0  136.75   83.0      0.0  22.01      80.0   
4432  9998212    173.0  46.0  141.00   89.0      0.0  25.60      67.0   
4433  9999312    196.0  46.0  138.00   86.0     20.0  24.19      90.0   

      GLUCOSE  HDLC   LDLC  SEX  educ  CVD  AGE_1  AGE_2  AGE_3  CURSMOKE_1  \
0        84.5  39.5  173.0    0   4.0    1     39      0     52           0   
1        76.0  47.0  165.0    1   2.0    0     46     52     58           0   
2        78.5  47.5  175.5    0   1.0    0     48     54      0           1   
3        96.0  46.5  178.5    1   3.0    1     61     67      0           1   
4        80.0  48.0  178.0    1   3.0    0     46     51     58           1   
...       ...   ...    ...  ...   ...  ...    ...    ...    ...         ...   
4429     86.0  47.0  176.0    1   2.0    0     48     54     60           1   
4430     78.0  49.0  170.0    1   1.0    0     44     50     56           1   
4431     93.0  47.0  173.0    1   2.0    1     52     58      0           0   
4432     72.0  47.0  164.0    0   3.0    0     40     46     52           0   
4433     80.0  48.0  168.0    1   3.0    0     39     46     50           1   

      CURSMOKE_2  CURSMOKE_3  DIABETES_1  DIABETES_2  DIABETES_3  BPMEDS_1  \
0              0           0           0           0           0       0.0   
1              0           0           0           0           0       0.0   
2              1           0           0           0           0       0.0   
3              1           0           0           0           0       0.0   
4              1           1           0           0           0       0.0   
...          ...         ...         ...         ...         ...       ...   
4429           1           1           0           0           0       0.0   
4430           1           1           0           0           0       0.0   
4431           0           0           0           0           0       0.0   
4432           0           0           0           0           0       0.0   
4433           1           1           0           0           0       0.0   

      BPMEDS_2  BPMEDS_3  PREVCHD_1  PREVCHD_2  PREVCHD_3  PREVAP_1  PREVAP_2  \
0          0.0       0.0          0          0          0         0         0   
1          0.0       0.0          0          0          0         0         0   
2          0.0       0.0          0          0          0         0         0   
3          0.0       0.0          0          0          0         0         0   
4          0.0       0.0          0          0          0         0         0   
...        ...       ...        ...        ...        ...       ...       ...   
4429       0.0       0.0          0          0          0         0         0   
4430       0.0       1.0          0          0          0         0         0   
4431       1.0       0.0          0          0          0         0         0   
4432       0.0       0.0          0          0          0         0         0   
4433       0.0       0.0          0          0          0         0         0   

      PREVAP_3  PREVMI_1  PREVMI_2  PREVMI_3  PREVSTRK_1  PREVSTRK_2  \
0            0         0         0         0           0           0   
1            0         0         0         0           0           0   
2            0         0         0         0           

<div class="alert alert-block alert-info"> 
<b>Comments</b><br>
- We have created additional columns.<br>
- We combined everything into one common final dataset.<br>
</div>

**Reducing memory  usage.**<br>

In [35]:
dataset_preprocessed = reduce_mem_usage.reduce_mem_usage(framingham_final_merged)

Memory usage of dataframe is 1.49 MB
Memory usage after optimization is: 0.38 MB
Decreased by 74.7%


In [36]:
dataset_preprocessed.sort_index()

RANDID  TOTCHOL   AGE   SYSBP  DIABP  CIGPDAY        BMI  HEARTRTE  \
0        2448    202.0  45.5  113.50   68.0      0.0  26.969999      74.5   
1        6238    250.0  52.0  108.00   69.5      0.0  28.730000      80.0   
2        9428    264.0  51.0  134.25   84.5     25.0  25.340000      75.0   
3       10552    228.5  64.0  166.50  102.0     25.0  29.379999      62.5   
4       11252    285.0  51.0  130.00   84.0     30.0  23.480000      85.0   
...       ...      ...   ...     ...    ...      ...        ...       ...   
4429  9990894    248.0  54.0  145.00   89.0     20.0  22.000000      80.0   
4430  9993179    236.0  50.0  131.00   87.0     25.0  21.219999      86.0   
4431  9995546    267.0  55.0  136.75   83.0      0.0  22.010000      80.0   
4432  9998212    173.0  46.0  141.00   89.0      0.0  25.600000      67.0   
4433  9999312    196.0  46.0  138.00   86.0     20.0  24.190001      90.0   

      GLUCOSE  HDLC   LDLC  SEX  educ  CVD  AGE_1  AGE_2  AGE_3  CURSMOKE_1  \
0        84.5  39.5  173.0    0   4.0    1     39      0     52           0   
1        76.0  47.0  165.0    1   2.0    0     46     52     58           0   
2        78.5  47.5  175.5    0   1.0    0     48     54      0           1   
3        96.0  46.5  178.5    1   3.0    1     61     67      0           1   
4        80.0  48.0  178.0    1   3.0    0     46     51     58           1   
...       ...   ...    ...  ...   ...  ...    ...    ...    ...         ...   
4429     86.0  47.0  176.0    1   2.0    0     48     54     60           1   
4430     78.0  49.0  170.0    1   1.0    0     44     50     56           1   
4431     93.0  47.0  173.0    1   2.0    1     52     58      0           0   
4432     72.0  47.0  164.0    0   3.0    0     40     46     52           0   
4433     80.0  48.0  168.0    1   3.0    0     39     46     50           1   

      CURSMOKE_2  CURSMOKE_3  DIABETES_1  DIABETES_2  DIABETES_3  BPMEDS_1  \
0              0           0           0           0           0       0.0   
1              0           0           0           0           0       0.0   
2              1           0           0           0           0       0.0   
3              1           0           0           0           0       0.0   
4              1           1           0           0           0       0.0   
...          ...         ...         ...         ...         ...       ...   
4429           1           1           0           0           0       0.0   
4430           1           1           0           0           0       0.0   
4431           0           0           0           0           0       0.0   
4432           0           0           0           0           0       0.0   
4433           1           1           0           0           0       0.0   

      BPMEDS_2  BPMEDS_3  PREVCHD_1  PREVCHD_2  PREVCHD_3  PREVAP_1  PREVAP_2  \
0          0.0       0.0          0          0          0         0         0   
1          0.0       0.0          0          0          0         0         0   
2          0.0       0.0          0          0          0         0         0   
3          0.0       0.0          0          0          0         0         0   
4          0.0       0.0          0          0          0         0         0   
...        ...       ...        ...        ...        ...       ...       ...   
4429       0.0       0.0          0          0          0         0         0   
4430       0.0       1.0          0          0          0         0         0   
4431       1.0       0.0          0          0          0         0         0   
4432       0.0       0.0          0          0          0         0         0   
4433       0.0       0.0          0          0          0         0         0   

      PREVAP_3  PREVMI_1  PREVMI_2  PREVMI_3  PREVSTRK_1  PREVSTRK_2  \
0            0         0         0         0           0           0   
1            0         0         0         0           0           0   
2      

**Saving the preprocessed dataset.**<br>

In [37]:
folder_path = 'csv/'
file_name = 'preprocessed_dataset.csv'
dataset_preprocessed.to_csv(folder_path + file_name, index=False)